core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -66,140 +72,221 @@ class LogManager(val logDirs: Array[File],
   private val logs = new Pool[TopicPartition, Log]()
   private val logsToBeDeleted = new LinkedBlockingQueue[Log]()
 
-  createAndValidateLogDirs(logDirs)
-  private val dirLocks = lockLogDirs(logDirs)
-  private val recoveryPointCheckpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, RecoveryPointCheckpointFile)))).toMap
-  private val logStartOffsetCheckpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, LogStartOffsetCheckpointFile)))).toMap
+  val liveLogDirs: ArrayBuffer[File] = createAndValidateLogDirs(logDirs, initialOfflineDirs)
+  @volatile private var recoveryPointCheckpoints = liveLogDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, RecoveryPointCheckpointFile)))).toMap
+  @volatile private var logStartOffsetCheckpoints = liveLogDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, LogStartOffsetCheckpointFile)))).toMap
+
+  private def offlineLogDirs = logDirs.filterNot(liveLogDirs.toSet)
+
   loadLogs()
 
+  private val dirLocks = lockLogDirs(liveLogDirs)
+
   // public, so we can access this from kafka.admin.DeleteTopicTest
   val cleaner: LogCleaner =
     if(cleanerConfig.enableCleaner)
-      new LogCleaner(cleanerConfig, logDirs, logs, time = time)
+      new LogCleaner(cleanerConfig, liveLogDirs, logs, this, time = time)
     else
       null
-  
+
+  val offlineLogDirectoryCount = newGauge(
```

Comment:
Previously I don't think it is needed because if user wants to know which specific log directories are offline in addition to the offline log directory count, it probably means they want to fix the problem and they will login the machine anyway. In this case they can find the offline log directory name in the log. And user probably wants to fix the problem because the log itself gets deleted from the cluster because they need to see the exception trace to debug the problem.

But it doesn't hurt to have extra metrics. I have updated the patch with the following code:

```
  for (dir <- logDirs) {
    newGauge(
      "OfflineLogDirectoryCount",
      new Gauge[Int] {
        def value = if (_liveLogDirs.contains(dir)) 0 else 1
      },
      Map("logDirectory" -> dir.getAbsolutePath)
    )
  }
```